mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraAttachmentDAOV2.java

```diff
@@ -0,0 +1,132 @@
+
+import org.apache.james.backends.cassandra.utils.CassandraAsyncExecutor;
+import org.apache.james.mailbox.cassandra.ids.BlobId;
+import org.apache.james.mailbox.cassandra.table.CassandraAttachmentTable;
+import org.apache.james.mailbox.model.Attachment;
+import org.apache.james.mailbox.model.AttachmentId;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.datastax.driver.core.PreparedStatement;
+import com.datastax.driver.core.Row;
+import com.datastax.driver.core.Session;
+import com.google.common.base.Preconditions;
+
+public class CassandraAttachmentDAOV2 {
+    private static final Logger LOGGER = LoggerFactory.getLogger(CassandraAttachmentMapper.class);
+    private static final boolean NO_LOG_IF_EMPTY = false;
+
+    private final CassandraAsyncExecutor cassandraAsyncExecutor;
+    private final PreparedStatement insertStatement;
+    private final CassandraBlobsDAO blobsDAO;
+    private final PreparedStatement selectStatement;
+
+    @Inject
+    public CassandraAttachmentDAOV2(Session session, CassandraBlobsDAO blobsDAO) {
+        this.cassandraAsyncExecutor = new CassandraAsyncExecutor(session);
+        this.blobsDAO = blobsDAO;
+
+        this.selectStatement = prepareSelect(session);
+        this.insertStatement = prepareInsert(session);
+    }
+
+    private PreparedStatement prepareInsert(Session session) {
+        return session.prepare(
+            insertInto(TABLE_NAME)
+                .value(ID, bindMarker(ID))
+                .value(BLOB_ID, bindMarker(BLOB_ID))
+                .value(TYPE, bindMarker(TYPE))
+                .value(SIZE, bindMarker(SIZE)));
+    }
+
+    private PreparedStatement prepareSelect(Session session) {
+        return session.prepare(select(FIELDS)
+            .from(TABLE_NAME)
+            .where(eq(ID, bindMarker(ID))));
+    }
+
+    public CompletableFuture<Optional<Attachment>> getAttachment(AttachmentId attachmentId) {
+        return getAttachment(attachmentId, NO_LOG_IF_EMPTY);
+    }
+
+    public CompletableFuture<Optional<Attachment>> getAttachment(AttachmentId attachmentId, boolean logIfEmpty) {
+        Preconditions.checkArgument(attachmentId != null);
+        return cassandraAsyncExecutor.executeSingleRow(
+            selectStatement.bind()
+                .setString(CassandraAttachmentTable.ID, attachmentId.getId()))
+            .thenCompose(this::attachment)
+            .thenApply(optional -> logNotFound(attachmentId, logIfEmpty, optional));
+    }
+
+    private Optional<Attachment> logNotFound(AttachmentId attachmentId, boolean logIfEmpty, Optional<Attachment> optional) {
+        if (!optional.isPresent() && logIfEmpty) {
+            LOGGER.warn("Failed retrieving attachment {}", attachmentId);
+        }
+        return optional;
+    }
+
+    public CompletableFuture<Void> storeAttachment(Attachment attachment) {
+        return blobsDAO.save(attachment.getBytes())
+            .thenApply(Optional::get) // attachment payload is never null
```

Comment:
you mean it's never empty ? Then why is it optional then ?

Reply:
BlobDAO supports saving **null** blobs. But attachments ensure there blobs are never null. Hence optional unboxing is safe here.

Reply:
I don't agree, if you need an API that guarantees not returning null/empty, add it instead of using that one.